### ISSUE TITLE
[BUG] `global` can overflow stack?

### ISSUE THREAD

**Issue by mattip** [cython/cython#4272]
I am trying to debug a strange crash in [scipy's sobel.pyx](https://github.com/scipy/scipy/blob/master/scipy/stats/_sobol.pyx) on pypy. I think (am not 100% sure) that the use of this paradigm is causing problems
```
DEF MAXDIM = 21201  # max number of dimensions
DEF MAXDEG = 18  # max polynomial degree
cdef int poly[MAXDIM]
cdef int vinit[MAXDIM][MAXDEG]
...
def initialize_direction_numbers():
    global is_initialized, poly, vinit
    if not is_initialized:
        dns = np.load(os.path.join(os.path.dirname(__file__), "_sobol_direction_numbers.npz"))
        poly = dns["poly"]
        vinit = dns["vinit"]
        is_initialized = True
```

This turns into C code like
```
static int __pyx_v_5scipy_5stats_6_sobol_poly[0x52D1];
static int __pyx_v_5scipy_5stats_6_sobol_vinit[0x52D1][18];
...

static PyObject *__pyx_pf_5scipy_5stats_6_sobol_initialize_direction_numbers(CYTHON_UNUSED PyObject *__pyx_self) {
  ....
  int __pyx_t_11[0x52D1];
  int __pyx_t_12[0x52D1][18];
 ....
    if (unlikely(__Pyx_carray_from_py_int___5b_18_5d_(__pyx_t_2, __pyx_t_12, 0x52D1) < 0)) __PYX_ERR(0, 98, __pyx_L1_error)
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (unlikely((0x52D1) != (0x52D1))) {
      PyErr_Format(PyExc_ValueError, "Assignment to slice of wrong length, expected %" CYTHON_FORMAT_SSIZE_T "d, got %" CYTHON_FORMAT_SSIZE_T "d", (Py_ssize_t)(0x52D1), (Py_ssize_t)(0x52D1));
      __PYX_ERR(0, 98, __pyx_L1_error)
    }
    memcpy(&(__pyx_v_5scipy_5stats_6_sobol_vinit[0]), __pyx_t_12, sizeof(__pyx_v_5scipy_5stats_6_sobol_vinit[0]) * (0x52D1));


```

So at the beginning of the function there is a very large array. Is it reasonable that this is causing a stack overflow? I see very strange phenomenon that point to that direction. Perhaps PyPy's C headers are preventing inlining the static function?

Maybe this is more applicable to the mailing list, but I am opening it as an issue since one alternative would be to avoid doubly statically allocating  the global 2d-array and memcpying it.

**Comment by da-woods** [comment]
> Perhaps PyPy's C headers are preventing inlining the static function?

I don't think a def function implementation would ever be meaningfully inlined (with/without PyPy).

>  I am opening it as an issue since one alternative would be to avoid doubly statically allocating the global 2d-array and memcpying it.

The reason it's being statically allocated then memcpyed is that `__Pyx_carray_from_py_int___5b_18_5d_` might throw (presumably because it's been passed data types that don't match). If it wrote directly into the global then this could leave it in a half-written state.

I'd think the solution (if we did want to fix it) would be to limit the size of C arrays we were prepared to create on the stack and make the temp an allocated variable instead.

**Comment by da-woods** [comment]
I suspect there isn't anything special about PyPy here - it's probably on the borderline of overflowing the stack in CPython too, but is just on the right side of the line.

**Comment by mattip** [comment]
> make the temp an allocated variable

Is there a simply change scipy could make to the code to avoid this problem?

**Comment by da-woods** [comment]
> Is there a simply change scipy could make to the code to avoid this problem?

probably to do the copy element-by-element itself:

```
dnspoly = dns["poly"]
for i in range(len(dnspoly)):
    poly[i] = dnspoly[i]
```
(and the same for `vint`)

**Comment by mattip** [comment]
Thanks. Feel free to close this or modify the topic to better suit any changes Cython might need to warn/prevent this.

**Comment by da-woods** [comment]
>  Feel free to close this or modify the topic to better suit any changes Cython might need to warn/prevent this.

Yeah - I'm not really sure what the right thing to do is. The fact that it's a huge hidden temp that the user has no reason to expect makes me think that it deserves a warning possibly? I think changing the code generation to avoid this error would be too complicated though

**Comment by mattip** [comment]
People tend to ignore warnings. I "fixed" the scipy code to avoid it, so I will close this. Thanks for the help.

**Comment by peterbell10** [comment]
Why close this? This bug is silent and results in crashes for reasons that are completely invisible to the Cython code.

If managing two versions of codegen would be too difficult then how about having a helper macro that either does stack allocation or `malloc` depending on the size.
```
#define SMALL_SIZE 8192
#define _Pyx_alloc_temp(size) (size <= SMALL_SIZE ? alloca(size) : malloc(size))
#define _Pyx_free_temp(ptr, size) if (size > SMALL_SIZE) { free(ptr); }

int* tmp = (int*)_Pyx_alloc_temp(0x52D1 * 18 * sizeof(int))
if (!tmp) {
    // Handle error
}
// Use tmp ...
_Pyx_free_temp(tmp, 0x52D1 * 18 * sizeof(int));

**Comment by robertwb** [comment]
I agree this is a subtle error that would be good to avoid. I think we're allocating the temp to make the reassignment of the global look atomic (it either happens completely, or not at all in case there was a bug). We could keep it open, but I don't know in what timeframe it would be addressed. (It's not just a matter of creating a macro that switches allocation type based on a size parameter, but refactoring the temp generation code to make sure it is always freed correctly and the usages to handle a pointer rather than value. Feasible, possibly easier if isolated to the buffer unpacking code, but nontrivial.)

**Comment by vytas7** [comment]
I've started experiencing a variation of this with `cdef int[0x10000]` in CPython3.14 and musllinux (probably it has very tight stack size limits):
```cython
# PERF(vytas): Cache hex characters lookup table
cdef int[0x10000] HEX_CHARS
HEX_CHARS[:] = build_hex_table()
```

`build_hex_table()` returns a Python list, but in order to assign, it allocates a `cdef int[0x10000]` copy on stack?

**Edit**: initializing via a one-by-one loop works fine. 👍
